libraries/CMakeLists.txt

```diff
@@ -1,16 +1,32 @@
-# Add build configuration for all 3rd party modules.
-add_subdirectory(${3RDPARTY_DIR})
-
-# Macro utility to clone the coreMQTT submodule.
-macro( clone_mqtt )
-        find_package( Git REQUIRED )
-        message( "Cloning submodule coreMQTT." )
-        execute_process( COMMAND rm -rf ${MODULES_DIR}/standard/coreMQTT
-                         COMMAND ${GIT_EXECUTABLE} submodule update --init --recursive libraries/standard/coreMQTT
+# Macro utility to clone a library and source-related submodules.
+macro( clone_path path )
+    find_package( Git REQUIRED )
+    # Attempt to clone submodules.
+    if( ${BUILD_CLONE_SUBMODULES} )
+        execute_process( COMMAND ${GIT_EXECUTABLE} submodule update --init --recursive -- ${path}
                          WORKING_DIRECTORY ${ROOT_DIR}
-                         RESULT_VARIABLE MQTT_CLONE_RESULT )
-
-        if( NOT ${MQTT_CLONE_RESULT} STREQUAL "0" )
-                message( FATAL_ERROR "Failed to clone coreMQTT submodule." )
+                         RESULT_VARIABLE LIB_CLONE_RESULT )
+        if( NOT ${LIB_CLONE_RESULT} STREQUAL "0" )
+            message( FATAL_ERROR "Failed to clone submodules." )
         endif()
+    endif()
 endmacro()
+
+clone_path( ${MODULES_DIR} )
+
+set( LIBRARY_SUBMODULES
+       ${MODULES_DIR}/standard/coreHTTP
+       ${MODULES_DIR}/standard/coreJSON
+       ${MODULES_DIR}/standard/coreMQTT
+       ${MODULES_DIR}/aws/device-defender-for-aws-iot-embedded-sdk
+       ${MODULES_DIR}/aws/device-shadow-for-aws-iot-embedded-sdk
+       ${MODULES_DIR}/aws/jobs-for-aws-iot-embedded-sdk )
+
+foreach(submodule_dir IN ITEMS ${LIBRARY_SUBMODULES})
+    if( NOT EXISTS ${submodule_dir}/.git )
+        message( FATAL_ERROR "The required submodule {} does not exist. Either clone it manually, or set BUILD_CLONE_SUBMODULES to 1 to automatically clone it during build." )
+    endif()
+endforeach()
```

Comment:
I don;t think this is required considering that the default behavior of CMake is to set the `BUILD_CLONE_SUBMODULES` flag to 1. Thus, this error would only occur when the customer is aware of this option, and has explicitly set it to `OFF`.